core/controller/src/main/scala/whisk/core/controller/actions/PrimitiveActions.scala

```diff
@@ -139,6 +199,356 @@ protected[actions] trait PrimitiveActions {
+            val response = ActivationResponse.whiskError(compositionActivationFailure)
+            completeAppActivation(user, session, response)
+          case Success(Left(activationId)) =>
+            session.logs += activationId
+            val response = ActivationResponse.whiskError(compositionActivationTimeout(activationId))
+            completeAppActivation(user, session, response)
+          case Success(Right(activation)) =>
+            session.logs += activation.activationId
+            session.duration += activation.duration.getOrElse(
+              activation.end.toEpochMilli - activation.start.toEpochMilli)
+            activation.annotations.get("limits") map { limitsAnnotation =>
+              limitsAnnotation.asJsObject.getFields("memory") match {
+                case Seq(JsNumber(memory)) =>
+                  session.maxMemory = Math.max(session.maxMemory, memory.toInt)
+              }
+            }
+
+            // invoke continuation
+            invokeConductor(user, Some(activation.resultAsJson), session)
+        }
+    }
+  }
+
+  /**
+   * Creates an activation for a composition and writes it back to the datastore.
+   * Completes the associated promise if any.
+   * Resumes caller if any.
+   */
+  private def completeAppActivation(user: Identity, session: Session, response: ActivationResponse)(
+    implicit transid: TransactionId): Unit = {
+
+    // compute max memory
+    val sequenceLimits = Parameters(
+      WhiskActivation.limitsAnnotation,
+      ActionLimits(session.action.limits.timeout, MemoryLimit(session.maxMemory MB), session.action.limits.logs).toJson)
+
+    // set causedBy if not topmost
+    val causedBy = if (session.cause.isDefined) {
+      Some(Parameters(WhiskActivation.causedByAnnotation, JsString(Exec.SEQUENCE)))
+    } else {
+      None
+    }
+
+    val end = Instant.now(Clock.systemUTC())
+
+    // create the whisk activation
+    val activation = WhiskActivation(
+      namespace = user.namespace.toPath,
+      name = session.action.name,
+      user.subject,
+      activationId = session.activationId,
+      start = session.start,
+      end = end,
+      cause = session.cause,
+      response = response,
+      logs = ActivationLogs(session.logs.map(_.asString).toVector),
+      version = session.action.version,
+      publish = false,
+      annotations = Parameters(WhiskActivation.topmostAnnotation, JsBoolean(!session.cause.isDefined)) ++
+        Parameters(WhiskActivation.pathAnnotation, JsString(session.action.fullyQualifiedName(false).asString)) ++
+        Parameters(WhiskActivation.kindAnnotation, JsString(Exec.SEQUENCE)) ++
+        Parameters(WhiskActivation.conductorAnnotation, JsBoolean(true)) ++
+        causedBy ++
+        sequenceLimits,
+      duration = Some(session.duration))
+
+    // complete the promise if any
+    session.result.map { _.success(Right(activation)) }
+
+    logging.info(this, s"recording activation '${activation.activationId}'")
```

Comment:
probably unnecessary - the entity store will log straddling the operation (now at debug level).